cuqi/experimental/mcmc/_gibbs.py

```diff
@@ -0,0 +1,258 @@
+
+        # Store target and allow conditioning to reduce to a single density
+        self.target = target() # Create a copy of target distribution (to avoid modifying the original)
+
+        # Store sampler instances (again as a copy to avoid modifying the original)
+        self.samplers = sampling_strategy.copy()
+
+        # Store number of sampling steps for each parameter
+        self.sampling_steps = sampling_steps
+
+        # Store parameter names
+        self.par_names = self.target.get_parameter_names()
+
+        # Initialize sampler (after target is set)
+        self._initialize()
+
+    def _initialize(self):
+        """ Initialize sampler """
+
+        # Initial points
+        self.current_samples = self._get_initial_points()
+
+        # Initialize sampling steps
+        self._initialize_sampling_steps()
+
+        # Allocate samples
+        self._allocate_samples()
+
+        # Set targets
+        self._set_targets()
+
+        # Initialize the samplers
+        self._initialize_samplers() 
+
+        # Run over pre-sample methods for samplers that have it
+        # TODO. Some samplers (NUTS) seem to require to run _pre_warmup before _pre_sample
+        # This is not ideal and should be fixed in the future
+        for sampler in self.samplers.values():
+            self._pre_warmup_and_pre_sample_sampler(sampler)
+
+        # Validate all targets for samplers.
+        self.validate_targets()
+
+    # ------------ Public methods ------------
+    def validate_targets(self):
+        """ Validate each of the conditional targets used in the Gibbs steps """
+        if not isinstance(self.target, JointDistribution):
+            raise ValueError('Target distribution must be a JointDistribution.')
+        for sampler in self.samplers.values():
+            sampler.validate_target()
+
+    def sample(self, Ns) -> 'GibbsNew':
+        """ Sample from the joint distribution using Gibbs sampling """
+        for _ in progressbar(range(Ns)):
+            self.step()
+            self._store_samples()
+
+    def warmup(self, Nb) -> 'GibbsNew':
+        """ Warmup (tune) the Gibbs sampler """
+        for idx in progressbar(range(Nb)):
+            self.step()
+            self.tune(idx)
+            self._store_samples()
+
+    def get_samples(self) -> Dict[str, Samples]:
+        samples_object = {}
+        for par_name in self.par_names:
+            samples_array = np.array(self.samples[par_name]).T
+            samples_object[par_name] = Samples(samples_array, self.target.get_density(par_name).geometry)
+        return samples_object
+    
+    def step(self):
+        """ Sequentially go through all parameters and sample them conditionally on each other """
+
+        # Sample from each conditional distribution
+        for par_name in self.par_names:
+
+            # Set targets (TODO: This is inefficient. Instead we should only update the target for the current parameter)
+            self._set_targets()
```

Comment:
Does it make sense to utilize individual sampler state in this case? To save the state before set_target and then set the state of the sampler after setting the target (instead of setting the initial point and scale, as other samplers might have other parameters that determine the tuning, e.g. NUTS).

Reply:
It is a bit more complicated then that I would say. In particular we need to make sure that changing the target in each sub-sampler is carried out correctly. I left this for another issue (and PR) #427 for that reason.

Reply:
sounds good!